File: benchmark/2026-09-17-7677581(latest)/standard/rubrics/probes/CONC-1/zig/conc1.zig
// CONC-1, Zig. Worker mechanism: std.Thread (Zig standard library).
const std = @import("std");

const N: i64 = 500000000;
const CHUNKS: usize = 4;
const SPAN: i64 = N / @as(i64, CHUNKS);

var partial: [CHUNKS]f64 = .{0.0} ** CHUNKS;

fn chunkSum(c: usize) f64 {
    var s: f64 = 0.0;
    const start: i64 = @as(i64, @intCast(c)) * SPAN;
    const end: i64 = start + SPAN;
    var i: i64 = start;
    while (i < end) : (i += 1) {
        const x = @sin(@as(f64, @floatFromInt(i)));
        s += x * x;
    }
    return s;
}

fn worker(w: usize, total_workers: usize) void {
    var c: usize = 0;
    while (c < CHUNKS) : (c += 1) {
        if (c % total_workers == w) partial[c] = chunkSum(c);
    }
}

pub fn main(init: std.process.Init.Minimal) !void {
    var it: std.process.Args.Iterator = .init(init.args);
    _ = it.skip();
    const w_count: usize = if (it.next()) |a| try std.fmt.parseInt(usize, a, 10) else 1;

    var threads: [8]std.Thread = undefined;
    var w: usize = 0;
    while (w < w_count) : (w += 1) {
        threads[w] = try std.Thread.spawn(.{}, worker, .{ w, w_count });
    }
    w = 0;
    while (w < w_count) : (w += 1) threads[w].join();

    var total: f64 = 0.0;
    var c: usize = 0;
    while (c < CHUNKS) : (c += 1) total = total + partial[c];
    std.debug.print("workers={d} result={d:.10}\n", .{ w_count, total });
}

File: benchmark/2026-09-17-7677581(latest)/standard/rubrics/probes/CONC-1/rust/conc1.rs
// CONC-1, Rust. Worker mechanism: std::thread::scope scoped threads (std).
const N: i64 = 500000000;
const CHUNKS: usize = 4;
const SPAN: i64 = N / CHUNKS as i64;

fn chunk_sum(c: usize) -> f64 {
    let mut s = 0.0f64;
    let start = c as i64 * SPAN;
    let end = start + SPAN;
    let mut i = start;
    while i < end { let x = (i as f64).sin(); s += x * x; i += 1; }
    s
}

fn main() {
    let w_count: usize = std::env::args().nth(1).unwrap_or_else(|| "1".to_string())
        .parse().unwrap();
    let mut partial = vec![0.0f64; CHUNKS];
    let results: Vec<Vec<(usize, f64)>> = std::thread::scope(|scope| {
        let mut hs = Vec::new();
        for w in 0..w_count {
            hs.push(scope.spawn(move || {
                let mut out = Vec::new();
                for c in 0..CHUNKS { if c % w_count == w { out.push((c, chunk_sum(c))); } }
                out
            }));
        }
        hs.into_iter().map(|h| h.join().unwrap()).collect()
    });
    for r in results { for (c, v) in r { partial[c] = v; } }
    let mut total = 0.0f64;
    for c in 0..CHUNKS { total = total + partial[c]; }
    println!("workers={} result={:.10}", w_count, total);
}
